crates/compilers/src/preprocessor/deps.rs

```diff
@@ -0,0 +1,275 @@
+    referenced_contracts: HashSet<u32>,
+}
+
+impl<'hir> BytecodeDependencyCollector<'hir> {
+    fn new(
+        source_map: &'hir SourceMap,
+        hir: &'hir Hir<'hir>,
+        src: &'hir str,
+        src_dir: &'hir PathBuf,
+    ) -> Self {
+        Self {
+            source_map,
+            hir,
+            src,
+            src_dir,
+            dependencies: vec![],
+            referenced_contracts: HashSet::default(),
+        }
+    }
+
+    /// Collects reference identified as bytecode dependency of analyzed contract.
+    /// Discards any reference that is not in project src directory (e.g. external
+    /// libraries or mock contracts that extend source contracts).
+    fn collect_dependency(&mut self, dependency: BytecodeDependency) {
+        let contract = Hir::contract(self.hir, ContractId::new(dependency.referenced_contract));
+        let source = Hir::source(self.hir, contract.source);
+        let FileName::Real(path) = &source.file.name else {
+            return;
+        };
+
+        if !path.starts_with(self.src_dir) {
+            let path = path.display();
+            trace!("ignore dependency {path}");
+            return;
+        }
+
+        self.referenced_contracts.insert(dependency.referenced_contract);
+        self.dependencies.push(dependency);
+    }
+}
+
+impl<'hir> Visit<'hir> for BytecodeDependencyCollector<'hir> {
+    type BreakValue = Never;
+
+    fn hir(&self) -> &'hir Hir<'hir> {
+        self.hir
+    }
+
+    fn visit_expr(&mut self, expr: &'hir Expr<'hir>) -> ControlFlow<Self::BreakValue> {
+        match &expr.kind {
+            ExprKind::New(ty) => {
+                if let TypeKind::Custom(item_id) = ty.kind {
+                    if let Some(contract_id) = item_id.as_contract() {
+                        let name_loc = SourceMapLocation::from_span(self.source_map, ty.span);
+                        let name = &self.src[name_loc.start..name_loc.end];
+                        // TODO: check if there's a better way to determine where constructor call
+                        // ends.
+                        let args_len = self.src[name_loc.end..].split_once(';').unwrap().0.len();
+                        self.collect_dependency(BytecodeDependency {
+                            kind: BytecodeDependencyKind::New(name.to_string(), args_len),
+                            loc: SourceMapLocation::from_span(
+                                self.source_map,
+                                Span::new(expr.span.lo(), expr.span.hi()),
+                            ),
+                            referenced_contract: contract_id.get(),
+                        });
+                    }
+                }
+            }
+            ExprKind::Member(member_expr, ident) => {
+                if ident.name.to_string() == "creationCode" {
+                    if let ExprKind::TypeCall(ty) = &member_expr.kind {
+                        if let TypeKind::Custom(contract_id) = &ty.kind {
```

Comment:
should move these checks up since they are cheaper than for `ident`

Reply:
hope I got this right, moved the kind checks above ident name check https://github.com/foundry-rs/compilers/pull/252/commits/f4a109d53bfe97bba287335404c2f5cc20e7dbf6